File: src/drscreen/preprocess/enhance.py
"""CLAHE, illumination normalization, and denoising. All mask-respecting:
pixels outside the FOV mask are never touched."""
import cv2
import numpy as np


def clahe_green(img: np.ndarray, mask: np.ndarray, clip_limit: float = 2.0, mode: str = "green") -> np.ndarray:
    """Apply CLAHE to either the green channel or L (of LAB), selectable via `mode`."""
    clahe = cv2.createCLAHE(clipLimit=clip_limit, tileGridSize=(8, 8))
    if mode == "green":
        out = img.copy()
        channel = img[:, :, 1]
        enhanced = clahe.apply(channel)
        out[:, :, 1] = np.where(mask, enhanced, channel)
        return out
    if mode == "lab":
        lab = cv2.cvtColor(img, cv2.COLOR_RGB2LAB)
        l_channel = lab[:, :, 0]
        lab[:, :, 0] = np.where(mask, clahe.apply(l_channel), l_channel)
        out = cv2.cvtColor(lab, cv2.COLOR_LAB2RGB)
        for c in range(3):
            out[:, :, c] = np.where(mask, out[:, :, c], img[:, :, c])
        return out
    raise ValueError(f"unknown clahe mode: {mode}")


def illumination_normalize(img: np.ndarray, mask: np.ndarray, sigma_frac: float = 0.05) -> np.ndarray:
    """Ben Graham style: subtract a heavy Gaussian blur, then rescale to uint8."""
    h, w = img.shape[:2]
    sigma = max(1.0, sigma_frac * max(h, w))
    blurred = cv2.GaussianBlur(img, (0, 0), sigmaX=sigma, sigmaY=sigma)
    diff = img.astype(np.float32) - blurred.astype(np.float32)
    normalized = (diff - diff.min()) / (diff.max() - diff.min() + 1e-8) * 255.0
    normalized = normalized.astype(np.uint8)
    out = img.copy()
    for c in range(img.shape[2]):
        out[:, :, c] = np.where(mask, normalized[:, :, c], img[:, :, c])
    return out


def denoise(img: np.ndarray, mask: np.ndarray, method: str = "bilateral") -> np.ndarray:
    """Denoise via bilateral filtering (fast) or non-local means (slower, higher quality)."""
    if method == "bilateral":
        denoised = cv2.bilateralFilter(img, d=5, sigmaColor=50, sigmaSpace=50)
    elif method == "nlm":
        denoised = cv2.fastNlMeansDenoisingColored(img, None, 6, 6, 7, 21)
    else:
        raise ValueError(f"unknown denoise method: {method}")
    out = img.copy()
    for c in range(img.shape[2]):
        out[:, :, c] = np.where(mask, denoised[:, :, c], img[:, :, c])
    return out


if __name__ == "__main__":
    import time

    from .retina import cached_fov_mask

    rng = np.random.default_rng(0)
    mask = cached_fov_mask(448)
    t0 = time.time()
    for _ in range(5):
        img = rng.integers(0, 255, (448, 448, 3), dtype=np.uint8)
        for fn, kwargs in [
            (clahe_green, {}),
            (clahe_green, {"mode": "lab"}),
            (illumination_normalize, {}),
            (denoise, {}),
        ]:
            out = fn(img, mask, **kwargs)
            assert out.shape == img.shape
            assert np.array_equal(out[~mask], img[~mask]), f"{fn.__name__}{kwargs} touched outside the FOV mask"
    elapsed = time.time() - t0
    print(f"enhance.py smoke test ok in {elapsed:.3f}s")
    assert elapsed < 30

File: matlab/enhance.m
function out = enhance(img, mask, method, varargin)
%ENHANCE CLAHE, illumination normalization, or denoising, restricted to
% pixels inside mask. Port of src/drscreen/preprocess/enhance.py. Named to
% match this file (enhance.m) for the same reason preprocess.m is -- see
% that file's header comment.
%
% NOT independently verified under Octave: Octave's image package (2.20.1,
% the version installed while building this) does not implement
% adapthisteq, so the 'clahe' path below could not be executed here. The
% others (imgaussfilt, bilateral filtering) use functions Octave does
% provide, but were not run against real data the way preprocess.m was --
% see matlab/README.md. All three are also not currently called from
% run_pipeline.m's live inference path, mirroring pipeline.py: this project's
% actual quality-gated inference never applies CLAHE/illumination-norm/
% denoise today, only preprocess_image's crop+resize (see pipeline.py's
% ScreeningPipeline.run). Ported anyway because AGENTS.md's Phase 6 plan
% asks for it explicitly.
%
%   out = enhance(img, mask, 'clahe', clipLimit)         % default clipLimit 2.0
%   out = enhance(img, mask, 'illumination', sigmaFrac)  % default sigmaFrac 0.05
%   out = enhance(img, mask, 'denoise')

switch method
    case 'clahe'
        clip_limit = 2.0;
        if ~isempty(varargin)
            clip_limit = varargin{1};
        end
        out = clahe_green(img, mask, clip_limit);
    case 'illumination'
        sigma_frac = 0.05;
        if ~isempty(varargin)
            sigma_frac = varargin{1};
        end
        out = illumination_normalize(img, mask, sigma_frac);
    case 'denoise'
        out = denoise_bilateral(img, mask);
    otherwise
        error('enhance:unknownMethod', 'unknown method: %s', method);
end
end


function out = clahe_green(img, mask, clip_limit)
% CLAHE on the green channel only, matching enhance.py's default mode="green".
out = img;
channel = img(:, :, 2);
enhanced = adapthisteq(channel, 'ClipLimit', clip_limit / 100, 'NumTiles', [8 8]);
% clipLimit/100: MATLAB's adapthisteq ClipLimit is normalized to [0, 1]
% (fraction of the tile histogram), unlike OpenCV's createCLAHE, whose
% clipLimit is an unnormalized histogram-bin count. This scaling is an
% approximation, not a derived equivalence -- unverified, no adapthisteq
% in the Octave environment this was written in. Re-tune once run for real.
out(:, :, 2) = uint8(double(enhanced) .* mask + double(channel) .* ~mask);
end


function out = illumination_normalize(img, mask, sigma_frac)
% Ben Graham style: subtract a heavy Gaussian blur, rescale to uint8.
[h, w, ~] = size(img);
sigma = max(1.0, sigma_frac * max(h, w));
blurred = imgaussfilt(img, sigma);
diff = double(img) - double(blurred);
normalized = (diff - min(diff(:))) / (max(diff(:)) - min(diff(:)) + 1e-8) * 255.0;
out = img;
for c = 1:size(img, 3)
    out(:, :, c) = uint8(normalized(:, :, c) .* mask + double(img(:, :, c)) .* ~mask);
end
end


function out = denoise_bilateral(img, mask)
% Bilateral filter, matching enhance.py's default method="bilateral".
% imbilatfilt ships in MATLAB's Image Processing Toolbox (R2018b+);
% Octave's image package does not have it, so this path is entirely
% unverified.
denoised = imbilatfilt(img, 2500, 5);  % degreeOfSmoothing, spatialSigma --
% chosen to approximate cv2.bilateralFilter(d=5, sigmaColor=50, sigmaSpace=50),
% not derived from an exact parameter mapping between the two libraries.
out = img;
for c = 1:size(img, 3)
    out(:, :, c) = uint8(double(denoised(:, :, c)) .* mask + double(img(:, :, c)) .* ~mask);
end
end
